File: 5_dp/이재영/3_도둑질.swift
//
//  3_도둑질.swift
//  https://school.programmers.co.kr/learn/courses/30/lessons/42897
//
//  Created by jerry on 9/8/26.
//

import Foundation
/*
 
 1. 처음 집 방문 O -> 마지막 집 방문 불가
 2. 처음 집 방문 X -> 마지막 집 방문 가능
 
 */
func solution(_ money: [Int]) -> Int {
    // 첫번째 집 방문
    let aCase = maxValue(money[0], 2, money.dropLast())
    // 첫번째 집 미방문
    let bCase = maxValue(0, 1, money)
    
    return max(aCase, bCase)
}

func maxValue(_ startValue: Int, _ startIndex: Int, _ money: [Int]) -> Int {
    var enable = startValue
    var disable = startValue
    
    for i in startIndex..<money.count {
        var newEnable = max(disable, enable)
        var newDisable = enable + money[i]
        
        enable = newEnable
        disable = newDisable
    }
    
    return max(enable, disable)
}

File: 5_dp/이재영/3_도둑질.java
class Solution {
    public int solution(int[] money) {
        // 배열 길이가 3 미만일 경우 예외 처리 (Swift 코드의 인덱스 접근 런타임 에러 방지)
        if (money.length < 3) return 0;
        
        // money.dropLast() 대신 검사할 끝 인덱스(money.length - 1)를 전달합니다.
        int aCase = maxValue(money[0], 2, money.length - 1, money);
        int bCase = maxValue(0, 1, money.length, money);
        
        return Math.max(aCase, bCase);
    }

    private int maxValue(int startValue, int startIndex, int endIndex, int[] money) {
        int enable = startValue;
        int disable = startValue;
        
        for (int i = startIndex; i < endIndex; i++) {
            int newEnable = Math.max(disable, enable);
            int newDisable = enable + money[i];
            
            enable = newEnable;
            disable = newDisable;
        }
        
        return Math.max(enable, disable);
    }
}
